### PR TITLE
Sessions tweaks

### LINKED ISSUE #2080: [master] Temp tokens (Webmail) expires on midnight
A webmail temp token contains the current date in it's HMAC. So it will expire on midnight, giving varying token lifetime between 0s and 24h:
https://github.com/Mailu/Mailu/blob/dbbfa44461c3b8a537e9fe6b56f4690417d3c8a2/core/admin/mailu/models.py#L651

Better approach:
* on token creation, get the curent date and time
    * use date and time as part of HMAC
    * append date+time in plaintext to the token
* on validation, use date+time from token
    * use to validate hmac
    * use to validate if the token is expired or not

### PR DESCRIPTION
**Pull request by nextgens**
## What type of PR?

bug-fix

## What does this PR do?

- Make all sessions permanent, introduce SESSION_TIMEOUT and PERMANENT_SESSION_LIFETIME.
- Prevent the creation of a session before there is a login attempt
- Ensure that webmail tokens are in sync with sessions

### Related issue(s)
- close #2080 

## Prerequisites
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [x] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

### CODE REVIEW

**Comment by nextgens** [comment]
bors try

**Comment by nextgens** [comment]
bors try

**Comment by Diman0** [comment]
With these values configured I receive an error after logging in
Mailu.env values:
```
SESSION_TIMEOUT=20
PERMANENT_SESSION_LIFETIME=60
```

Admin trace:
```
admin_1      | Traceback (most recent call last):
admin_1      |   File "/usr/lib/python3.9/site-packages/flask/app.py", line 2072, in wsgi_app
admin_1      |     ctx.push()
admin_1      |   File "/usr/lib/python3.9/site-packages/flask/ctx.py", line 404, in push
admin_1      |     self.session = session_interface.open_session(self.app, self.request)
admin_1      |   File "/app/mailu/utils.py", line 379, in open_session
admin_1      |     return MailuSession(request.cookies.get(app.config['SESSION_COOKIE_NAME'], None), app)
admin_1      |   File "/app/mailu/utils.py", line 195, in __init__
admin_1      |     if parsed := self.app.session_config.parse_key(key, self.app):
admin_1      |   File "/app/mailu/utils.py", line 360, in parse_key
admin_1      |     if not created <= now <= created + app.config['PERMANENT_SESSION_LIFETIME']:
admin_1      | TypeError: unsupported operand type(s) for +: 'int' and 'str'
admin_1      | [2021-12-20 20:48:43,523] ERROR in app: Request finalizing failed with an error while handling an error
admin_1      | Traceback (most recent call last):
admin_1      |   File "/usr/lib/python3.9/site-packages/flask/app.py", line 2072, in wsgi_app
admin_1      |     ctx.push()
admin_1      |   File "/usr/lib/python3.9/site-packages/flask/ctx.py", line 404, in push
admin_1      |     self.session = session_interface.open_session(self.app, self.request)
admin_1      |   File "/app/mailu/utils.py", line 379, in open_session
admin_1      |     return MailuSession(request.cookies.get(app.config['SESSION_COOKIE_NAME'], None), app)
admin_1      |   File "/app/mailu/utils.py", line 195, in __init__
admin_1      |     if parsed := self.app.session_config.parse_key(key, self.app):
admin_1      |   File "/app/mailu/utils.py", line 360, in parse_key
admin_1      |     if not created <= now <= created + app.config['PERMANENT_SESSION_LIFETIME']:
admin_1      | TypeError: unsupported operand type(s) for +: 'int' and 'str'
admin_1      | 
admin_1      | During handling of the above exception, another exception occurred:
admin_1      | 
admin_1      | Traceback (most recent call last):
admin_1      |   File "/usr/lib/python3.9/site-packages/flask/app.py", line 1540, in finalize_request
admin_1      |     response = self.process_response(response)
admin_1      |   File "/usr/lib/python3.9/site-packages/flask/app.py", line 1885, in process_response
admin_1      |     response = self.ensure_sync(func)(response)
admin_1      |   File "/usr/lib/python3.9/site-packages/flask_login/login_manager.py", line 399, in _update_remember_cookie
admin_1      |     if '_remember' not in session and \
admin_1      | TypeError: argument of type 'NoneType' is not iterable

```

File: core/admin/mailu/utils.py [hunk 1]
```diff
@@ -363,13 +353,12 @@ def parse_key(self, key, app=None, validate=False, now=None):
...
-            if now is None:
-                now = int(time.time())
-            created = int.from_bytes(created, byteorder='big')
-            if not created < now < created + app.permanent_session_lifetime.total_seconds():
-                return None
+        # validate creation time
+        if now is None:
+            now = int(time.time())
+        created = int.from_bytes(created, byteorder='big')
+        if not created <= now <= created + app.config['PERMANENT_SESSION_LIFETIME']:
```
Comment (Diman0) [inline core/admin/mailu/utils.py:360]: See stack trace. i think app.config['PERMANENT_SESSION_LIFETIME'] must be cast  to a string.
Something like int( app.config['PERMANENT_SESSION_LIFETIME'] )?

**Review by Diman0 — CHANGES_REQUESTED** [review]
See previous comment

File: webmails/roundcube/config.inc.php [hunk 2]
```diff
@@ -13,6 +13,7 @@
 $config['zipdownload_selection'] = true;
 $config['enable_spellcheck'] = true;
 $config['spellcheck_engine'] = 'pspell';
+$config['session_lifetime'] = {{ SESSION_TIMEOUT // 60 }};
```
Comment (Diman0) [inline webmails/roundcube/config.inc.php:16]: This setting is missing in the roundcube deployed image. Perhaps the '// 60'  inflicts with the jinja scripting?
I will try removing it and  check if the config file is properly created.

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
Lessons learned. In jinja scripting to set a number use | int. E.g.
`{{ VARIABLE | int }}`
The Jina command expect the context (environment). If you change an environment variable, don't forget to update the context that you are going to pass to the jinja command.
In case someone picks a stupid low value for the timeout, the timeout value will be set to 1 minute in roundcube.

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
For unknown reason logging in docker fails in the CI workflow. Maybe there is a temporary issue with docker.
I will start the CI tomorrow morning again.

**Comment by Diman0** [comment]
bors try

**Comment by nextgens** [comment]
bors cancel

bors try

**Comment by nextgens** [comment]
bors cancel

**Comment by nextgens** [comment]
bors try

**Review by Diman0 — APPROVED** [review]
lgtm

File: core/admin/mailu/utils.py [hunk 3]
```diff
@@ -408,23 +397,12 @@ def save_session(self, app, session, response):
...
-            set_cookie |= session.save()
-
-        # set cookie on refreshed permanent sessions
-        if need_refresh and session.permanent:
-            set_cookie = True
-
-        # set or update cookie if necessary
-        if set_cookie:
+        # save session and update cookie if necessary
+        if session.save():
```
Comment (ghostwheel42) [inline core/admin/mailu/utils.py:401]: Why did you revert back to "save&refresh on every request"?
I did this intentionally after SESSION_TIMEOUT/2 (half the session time).

File: core/admin/mailu/utils.py [hunk 4]
```diff
@@ -498,3 +476,24 @@ def cleaner():
 
 cleaned = Value('i', False)
 session = MailuSessionExtension()
+
+# this is used by the webmail to authenticate IMAP/SMTP
+def verify_temp_token(email, token):
+    try:
+        if token.startswith('token-'):
+            sessid = app.session_store.get(token)
+            if sessid:
```
Comment (ghostwheel42) [inline core/admin/mailu/utils.py:485]: can be written as
`if sessid := app.session_store.get(token):`

File: webmails/roundcube/start.py [hunk 5]
```diff
@@ -62,6 +62,10 @@
 # add overrides
 context["INCLUDES"] = sorted(inc for inc in os.listdir("/overrides") if inc.endswith(".inc")) if os.path.isdir("/overrides") else []
 
+# calculate variables for config file
+env["SESSION_TIMEOUT_MINUTES"] = str(int(env.get("SESSION_TIMEOUT", "3600")) // 60 ) if int(env.get("SESSION_TIMEOUT", "3600")) >= 60 else "1"
+context.update(env)
```
Comment (ghostwheel42) [inline webmails/roundcube/start.py:67]: `context.update(env)`should not be called here.
Just set the converted value like this:
`context"SESSION_TIMEOUT_MINUTES"] = str(int(env.get("SESSION_TIMEOUT", "3600")) // 60 ) if int(env.get("SESSION_TIMEOUT", "3600")) >= 60 else "1"`

File: core/admin/mailu/utils.py (hunk 4 above)
Comment (nextgens) [reply core/admin/mailu/utils.py:485]: That's c161a2c9

File: core/admin/mailu/utils.py [hunk 6]
```diff
@@ -274,6 +267,11 @@ def save(self):
         if self._sid is None:
             self._sid = self.app.session_config.gen_sid()
             set_cookie = True
+            if 'webmail_token' in self:
+                app.session_store.put(self['webmail_token'],
+                        self.sid,
+                        app.config['PERMANENT_SESSION_LIFETIME'],
+                )
```
[the remark this answers is not in the document]
Comment (nextgens) [reply core/admin/mailu/utils.py:274]: That's f0247a2f

File: core/admin/mailu/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nextgens) [reply core/admin/mailu/utils.py:360]: That's f0247a2f

File: webmails/roundcube/start.py (hunk 5 above)
Comment (nextgens) [reply webmails/roundcube/start.py:67]: That's 6d5926ef

**Pull request merged**